File: frontend/app/models/crm.ts
import type { ApiFetch } from './shared'

export interface Cliente {
    cliente_id: string
    nome: string
    cpf_cnpj: string
    telefone: string | null
    email: string | null
    bloqueado: boolean
    ativo: boolean
}

export interface ClienteDados {
    nome: string
    cpf_cnpj: string
    telefone: string | null
    email: string | null
}

export function listarClientes(apiFetch: ApiFetch) {
    return apiFetch<{ clientes: Cliente[] }>('/crm/clientes')
}

export function criarCliente(apiFetch: ApiFetch, dados: ClienteDados) {
    return apiFetch('/crm/clientes', { method: 'POST', body: dados })
}

export function atualizarCliente(apiFetch: ApiFetch, clienteId: string, dados: Omit<ClienteDados, 'cpf_cnpj'>) {
    return apiFetch(`/crm/clientes/${clienteId}`, { method: 'PUT', body: dados })
}

export function bloquearCliente(apiFetch: ApiFetch, clienteId: string, motivo: string) {
    return apiFetch(`/crm/clientes/${clienteId}/bloquear`, { method: 'POST', body: { motivo } })
}

export function desbloquearCliente(apiFetch: ApiFetch, clienteId: string) {
    return apiFetch(`/crm/clientes/${clienteId}/desbloquear`, { method: 'POST' })
}

export function alternarAtivoCliente(apiFetch: ApiFetch, clienteId: string, ativo: boolean) {
    return apiFetch(`/crm/clientes/${clienteId}/${ativo ? 'desativar' : 'reativar'}`, { method: 'POST' })
}

File: src/web/routes/crm.rs
use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use pharos_app::{dispatch, query_dispatch};
use serde_json::json;
use uuid::Uuid;

use crate::auth::{AuthUser, Role, Roles};
use crate::crm::application::commands::{
    AtualizarCliente, BloquearCliente, CadastrarCliente, DesativarCliente, DesbloquearCliente,
    ReativarCliente,
};
use crate::crm::application::queries::{BuscarCliente, ListarClientes};
use crate::error::AppError;
use crate::web::{error::ApiError, state::CrmState};

pub async fn listar(
    State(s): State<CrmState>,
    user: AuthUser,
) -> Result<Json<serde_json::Value>, ApiError> {
    user.exigir_qualquer_role(&[Role::Vendedor])?;
    let clientes = query_dispatch(&*s.crm, ListarClientes).await?;
    Ok(Json(json!({ "clientes": clientes })))
}

pub async fn buscar(
    State(s): State<CrmState>,
    user: AuthUser,
    Path(cliente_id): Path<Uuid>,
) -> Result<Json<serde_json::Value>, ApiError> {
    user.exigir_qualquer_role(&[Role::Vendedor])?;
    let cliente = query_dispatch(&*s.crm, BuscarCliente { cliente_id }).await?;
    cliente
        .map(|c| Json(json!(c)))
        .ok_or_else(|| AppError::NotFound.into())
}

pub async fn cadastrar(
    State(s): State<CrmState>,
    user: AuthUser,
    Json(cmd): Json<CadastrarCliente>,
) -> Result<(StatusCode, Json<serde_json::Value>), ApiError> {
    user.exigir_qualquer_role(&[Role::Vendedor])?;
    let id = dispatch(&*s.crm, cmd).await?;
    Ok((
        StatusCode::CREATED,
        Json(json!({ "cliente_id": id.to_string() })),
    ))
}

pub async fn atualizar(
    State(s): State<CrmState>,
    user: AuthUser,
    Path(cliente_id): Path<Uuid>,
    Json(mut cmd): Json<AtualizarCliente>,
) -> Result<StatusCode, ApiError> {
    user.exigir_qualquer_role(&[Role::Vendedor])?;
    cmd.cliente_id = cliente_id;
    dispatch(&*s.crm, cmd).await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn bloquear(
    State(s): State<CrmState>,
    user: AuthUser,
    Path(cliente_id): Path<Uuid>,
    Json(mut cmd): Json<BloquearCliente>,
) -> Result<StatusCode, ApiError> {
    user.exigir_role(Roles::ADMIN)?;
    cmd.cliente_id = cliente_id;
    dispatch(&*s.crm, cmd).await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn desbloquear(
    State(s): State<CrmState>,
    user: AuthUser,
    Path(cliente_id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    user.exigir_role(Roles::ADMIN)?;
    dispatch(&*s.crm, DesbloquearCliente { cliente_id }).await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn desativar(
    State(s): State<CrmState>,
    user: AuthUser,
    Path(cliente_id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    user.exigir_role(Roles::ADMIN)?;
    dispatch(&*s.crm, DesativarCliente { cliente_id }).await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn reativar(
    State(s): State<CrmState>,
    user: AuthUser,
    Path(cliente_id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    user.exigir_role(Roles::ADMIN)?;
    dispatch(&*s.crm, ReativarCliente { cliente_id }).await?;
    Ok(StatusCode::NO_CONTENT)
}
